Packet++/src/DoIpLayer.cpp

```diff
@@ -0,0 +1,214 @@
+		if (it != DoIpEnumToStringPayloadType.end())
+		{
+			return it->second;
+		}
+		else
+		{
+			return "Unknown Payload type";
+		}
+	}
+
+	uint32_t DoIpLayer::getPayloadLength() const
+	{
+		return htobe32(getDoIpHeader()->payloadLength);
+	}
+
+	void DoIpLayer::setPayloadength(uint32_t Payloadength)
+	{
+		getDoIpHeader()->payloadLength = be32toh(Payloadength);
+	}
+
+	bool DoIpLayer::resolveProtocolVersion() const
+	{
+		DoIpProtocolVersion version = getProtocolVersion();
+		uint8_t inVersion = getInvertProtocolVersion();
+		DoIpPayloadTypes type = getPayloadType();
+
+		// Idea is token from wireshark
+		if (!(version == DoIpProtocolVersion::version01Iso2010 || version == DoIpProtocolVersion::version02Iso2012 ||
+		      version == DoIpProtocolVersion::version03Iso2019 ||
+		      version == DoIpProtocolVersion::version04Iso2019_AMD1 ||
+		      (version == DoIpProtocolVersion::defaultVersion &&
+		       (type >= DoIpPayloadTypes::VEHICLE_IDENTIFICATION_REQUEST &&
+		        type <= DoIpPayloadTypes::VEHICLE_IDENTIFICATION_REQUEST_WITH_VIN))))
+		{
+			PCPP_LOG_ERROR("Invalid/unsupported DoIP version!");
+			return false;
+		}
+		if ((uint8_t)(version) != (uint8_t) ~(inVersion))
+		{
+			PCPP_LOG_ERROR("Version and invert version are not synchronised !");
+			return false;
+		}
+		return true;
+	}
+
+	bool DoIpLayer::resolvePayloadLength() const
+	{
+		uint32_t length = getPayloadLength();
+
+		if (m_DataLen < sizeof(doiphdr))
+		{
+			PCPP_LOG_ERROR("Payload length is smaller than the minimum header size");
+			return false;
+		}
+
+		if (length != (m_DataLen - sizeof(doiphdr)))
+		{
+			PCPP_LOG_ERROR("Payload length does not match expected size");
+			return false;
+		}
+
+		return true;
+	}
+	bool DoIpLayer::resolveLayer() const
+	{
+		// Validate the protocol version and payload length
+		if (!resolveProtocolVersion() || !resolvePayloadLength())
+		{
+			PCPP_LOG_ERROR("Failed to Parse DoIP layer");
+			return false;
+		}
+		return true;
+	}
+
+	std::string DoIpLayer::toString() const
+	{
+		std::stringstream os;
+		if (!resolveLayer())
+		{
+			os << "Malformed doip Packet";
+			return os.str();
+		}
+		DoIpProtocolVersion version = getProtocolVersion();
+		DoIpPayloadTypes type = getPayloadType();
+		uint32_t length = getPayloadLength();
+
+		os << "DOIP Layer:" << std::endl;
+		os << "Protocol Version: " << getProtocolVersionAsStr() << std::hex << " (0x" << unsigned((uint8_t)version)
+		   << ")" << std::endl;
+		os << "Payload Type: " << getPayloadTypeAsStr() << std::hex << " (0x" << std::setw(4) << std::setfill('0')
+		   << (uint16_t)type << ")" << std::endl;
+		os << std::dec << "Payload Length: " << length << std::endl;
+
+		return os.str();
+	}
+
+	void DoIpLayer::serializeData(uint8_t* dest, std::vector<uint8_t> data)
+	{
+		memcpy(dest, data.data(), data.size());
+	}
+
+	void DoIpLayer::initLayer()
+	{
+		m_DataLen = sizeof(doiphdr);
+		m_Protocol = DOIP;
+		m_Data = new uint8_t[m_DataLen];
+		memset(m_Data, 0, m_DataLen);
```

Comment:
This should also zero initialize the array.

```suggestion
		m_Data = new uint8_t[m_DataLen]{};
```